File: src/main/java/burp/ls/XMLSource.java
package burp.ls;

import java.io.StringWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.xpath.*;

import burp.reporting.BurpException;
import burp.reporting.Origin;
import burp.reporting.RmlError;
import burp.vocabularies.RER;
import org.apache.commons.io.IOUtils;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import burp.model.Iteration;
import burp.util.SimpleNamespaceContext;

public class XMLSource extends FileBasedLogicalSource {

    public HashMap<String, String> prefixMap;

    @Override
    public Iterator<Iteration> iterator() throws BurpException {
        try {
            if (iterations == null) {
                iterations = new ArrayList<>();

                String contents = Files.readString(Paths.get(getDecompressedFile()), encoding);

                DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
                if (prefixMap != null) {
                    // Required for prefix evaluation of XPath expression
                    builderFactory.setNamespaceAware(true);
                }
                DocumentBuilder builder = builderFactory.newDocumentBuilder();
                Document xmlDocument = builder.parse(IOUtils.toInputStream(contents, encoding));

                XPath xPath = XPathFactory.newInstance().newXPath();
                if (prefixMap != null) {
                    SimpleNamespaceContext namespaces = new SimpleNamespaceContext(prefixMap);
                    xPath.setNamespaceContext(namespaces);
                }

                NodeList nodes = (NodeList) xPath.compile(iterator).evaluate(xmlDocument, XPathConstants.NODESET);

                for (int i = 0; i < nodes.getLength(); i++) {
                    Node node = nodes.item(i);
                    iterations.add(new XMLIteration(node, nulls, prefixMap));
                }
            }
            return iterations.iterator();
        } catch (XPathExpressionException e) {
            throw new BurpException(new RmlError(e.getMessage(), iteratorOrigin, RER.ReferenceFormulationSyntaxError, e, Collections.emptyMap()));
        } catch (Exception e) {
            throw new BurpException(new RmlError(e.getMessage(), iteratorOrigin, RER.ReferenceFormulationExecutionError, e, Collections.emptyMap()));
        }
    }

}

File: src/main/java/burp/ls/XMLSource.kt
package burp.ls

import burp.model.Iteration
import burp.reporting.BurpException
import burp.reporting.Origin
import burp.reporting.RmlError
import burp.vocabularies.RER
import burp.vocabularies.RML
import net.sf.saxon.s9api.Processor
import net.sf.saxon.s9api.SaxonApiException
import net.sf.saxon.s9api.XPathCompiler
import org.apache.jena.rdf.model.Resource
import java.nio.file.Files
import java.nio.file.Paths
import javax.xml.transform.stream.StreamSource

class XMLSource : FileBasedLogicalSource() {
    var iterator: String? = null
    var iteratorOrigin: Origin? = null

    var prefixMap: Map<String, String>? = null

    private var xPathCompiler: XPathCompiler? = null

    @Throws(BurpException::class)
    override fun iterator(): Iterator<Iteration> {
        try {
            if (iterations == null) {
                iterations = mutableListOf()

                val xmlDocument = Files.newBufferedReader(Paths.get(getDecompressedFile()), encoding).use { reader ->
                    documentBuilder.build(StreamSource(reader))
                }

                xPathCompiler = processor.newXPathCompiler()
                if (prefixMap != null) {
                    for ((prefix, uri) in prefixMap!!) {
                        xPathCompiler!!.declareNamespace(prefix, uri)
                    }
                }

                val selector = xPathCompiler!!.compile(iterator).load()
                selector.contextItem = xmlDocument
                val nodes = selector.iterator()

                nodes.forEach { iterations!!.add(XMLIteration(it, nulls, xPathCompiler!!)) }
            }
            return iterations!!.iterator()
        } catch (e: SaxonApiException) {
            throw BurpException(
                RmlError(
                    e.message!!,
                    iteratorOrigin,
                    RER.ReferenceFormulationSyntaxError,
                    e,
                    mutableMapOf()
                )
            )
        } catch (e: Exception) {
            throw BurpException(
                RmlError(
                    e.message!!,
                    iteratorOrigin,
                    RER.ReferenceFormulationExecutionError,
                    e,
                    mutableMapOf()
                )
            )
        }
    }

    override var referenceFormulation: Resource
        get() = RML.XPath
        set(value) {}

    override fun buildReference(reference: String, origin: Origin): burp.model.Reference {
        return XMLReference(reference, origin)
    }

    companion object {
        val processor = Processor(false)
        val documentBuilder = processor.newDocumentBuilder()
    }
}

class XMLReference(reference: String?, origin: Origin) : burp.model.Reference(reference, origin) {
    override fun getValues(i: Iteration): List<Any?> {
        require(i is XMLIteration)
        return try {
            val selector = i.xPathCompiler.compile(reference).load()
            selector.contextItem = i.node
            val nodes = selector.iterator()
            val l = mutableListOf<String>()
            nodes.forEach { l.add(it.stringValue) }
            l
        } catch (e: Exception) {
            throw BurpException(
                RmlError(
                    "Error executing XPath: $reference on node $i.",
                    origin,
                    RER.ReferenceFormulationExecutionError,
                    e
                )
            )
        }
    }
}

